File: aec/aec/report/export_opportunites_generalization2/export_opportunites_generalization2.py
# # Copyright (c) 2024, ds and contributors
# # For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data


def get_columns(filters):
    return [
        {
            "label": _("Member code"),
            "fieldname": "member",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 150,
        },
        {
            "label": _("countries"),
            "fieldname": "country",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("التكتلات"),
            "fieldname": "clusters",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("CIO Name"),
            "fieldname": "custom_name_of_the_cioowner_of_the_company",
            "fieldtype": "Data",
            "width": 200,
        },
        {
            "label": _("Committee Name"),
            "fieldname": "committee_name",
            "fieldtype": "Link",
            "options": "Committee",
            "width": 200,
        },
        {
            "label": _("Total Amount in EGP"),
            "fieldname": "total_amount_in_egp",
            "fieldtype": "Currency",
            "width": 150,
        },
        {
            "label": _("Total Amount in USD"),
            "fieldname": "total_amount_in_usd",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("current season"),
            "fieldname": "current_season",
            "fieldtype": "data",
            "width": 150,
        },
        {
            "label": _("Governorate code"),
            "fieldname": "territory_code",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("Governorate"),
            "fieldname": "territory",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("Membership status"),
            "fieldname": "custom_customer_status",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("CEO Mobile"),
            "fieldname": "customer_primary_contact",
            "fieldtype": "Data",
            "width": 200,
        },
        {"label": _("Email"), "fieldname": "email", "fieldtype": "Email", "width": 150},
    ]


def get_data(filters):
    conditions = []
    member = filters.get("member")
    country = filters.get("country")
    clusters = filters.get("clusters")
    custom_name_of_the_cioowner_of_the_company = filters.get(
        "custom_name_of_the_cioowner_of_the_company"
    )
    committee_name = filters.get("committee_name")
    territory = filters.get("territory")
    custom_customer_status = filters.get("custom_customer_status")
    if member:
        conditions.append("`tabCustomer`.`name` = %(member)s")
    else:
        conditions.append("`tabCustomer`.`name` IS NOT NULL")
    if country:
        conditions.append(
            "`tabVolume Of Member Exports`.`country_in_arabic` IN %(country_in_arabic)s"
        )
    if clusters:
        conditions.append("`tabCountries`.`geographical_clusters_name` IN %(clusters)s")
    if custom_name_of_the_cioowner_of_the_company:
        conditions.append(
            "`tabCustomer`.`custom_name_of_the_cioowner_of_the_company` = %(custom_name_of_the_cioowner_of_the_company)s"
        )
    if committee_name:
        conditions.append(
            "`tabCommittees you would like to join`.`committees` = %(committee_name)s"
        )
    if territory:
        conditions.append("`tabCustomer`.`territory` = %(territory)s")
    if custom_customer_status:
        conditions.append(
            "`tabCustomer`.`custom_customer_status` = %(custom_customer_status)s"
        )

    # Convert conditions list to a string for SQL query
    conditions_str = " AND ".join(conditions)
    if conditions_str:
        conditions_str = "WHERE " + conditions_str
    sql = f"""
        SELECT
            `tabCustomer`.`name` AS `member`,
            GROUP_CONCAT(DISTINCT `tabVolume Of Member Exports`.`country_in_arabic` SEPARATOR ', ') AS `country`,
            GROUP_CONCAT(DISTINCT `tabCountries`.`geographical_clusters_name` SEPARATOR ', ') AS `clusters`,
            `tabCustomer`.`custom_name_of_the_cioowner_of_the_company` AS `custom_name_of_the_cioowner_of_the_company`,
            GROUP_CONCAT(DISTINCT `tabCommittees you would like to join`.`committees` SEPARATOR ', ') AS `committee_name`,
            `tabVolume Of Member Exports for Three Years`.`total_amount_in_egp` AS `total_amount_in_egp`,
            CONCAT(FLOOR(`tabVolume Of Member Exports for Three Years`.`total_amount_in_usd`), ' $') AS `total_amount_in_usd`,
            CONCAT(YEAR(CURDATE()) - 2 , '-', YEAR(CURDATE()) - 1) AS `current_season`,
            `tabCustomer`.`territory` AS `territory_code`,
            `tabCustomer`.`governorate_name` AS `territory`,
            `tabCustomer`.`custom_ceo_mobile` AS `customer_primary_contact`,
            `tabCustomer`.`custom_customer_status` AS `custom_customer_status`,
            `tabCustomer`.`custom_email` AS `email`    
        FROM
            `tabVolume Of Member Exports`
        left JOIN `tabCustomer`
            ON `tabVolume Of Member Exports`.`tax__number` = `tabCustomer`.`tax_id`    
        LEFT JOIN `tabCommittees you would like to join`
            ON `tabCommittees you would like to join`.`parent` = `tabCustomer`.`name`
        LEFT JOIN `tabVolume Of Member Exports for Three Years`
            ON `tabVolume Of Member Exports for Three Years`.`parent` = `tabCustomer`.`name`

        Right JOIN `tabCountries`
            ON `tabCountries`.`name` = `tabVolume Of Member Exports`.`country_code`
        {conditions_str}
        GROUP BY `member`;
    """

    mydata = frappe.db.sql(
        sql,
        {
            "member": member,
            "country_in_arabic": country,
            "clusters": clusters,  # Pass clusters as a list
            "custom_name_of_the_cioowner_of_the_company": custom_name_of_the_cioowner_of_the_company,
            "committee_name": committee_name,
            "territory": territory,
            "custom_customer_status": custom_customer_status,
        },
        as_dict=True,
    )

    print(mydata)
    return mydata


# GROUP_CONCAT(DISTINCT `tabCountries`.`english_name` SEPARATOR ', ')AS country,
# 			GROUP_CONCAT(DISTINCT `tabCountries`.`geographical_clusters_name` SEPARATOR ', ') AS clusters,


#     inner JOIN `tabCountries`
# 			ON `tabCountries`.`name` = `tabVolume Of Member Exports`.`country_code`

File: aec/aec/report/export_opportunites_generalization2/export_opportunites_generalization2.js
// Copyright (c) 2024, ds and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Export Opportunites Generalization2"] = {
	filters: [
	  {
		fieldname: "member",
		label: __("Customer"),
		fieldtype: "Link",
		options: "Customer",
		width: "200px",
	  },
  
	  {
		fieldname: "committee_name",
		label: __("Please insert Committee"),
		fieldtype: "Link",
		options: "Committee",
		width: "350px",
	  },
	  {
		fieldname: "territory_code",
		label: __("Please Select Governorate"),
		fieldtype: "Link",
		options: "Territory",
		width: "350px",
	  },
  
	  {
		fieldname: "custom_customer_status",
		label: "Please select Membership Status",
		fieldtype: "Select",
		options: [
		  "",
		  "Requested",
		  "Requested From Website",
		  "Active",
		  "Inactive",
		  "Suspended",
		  "استيفاء بيانات",
		],
		default: "",
	  },
	  {
		fieldname: "name_of_the_cioowner_of_the_company",
		label: __("Please write CIO arabic"),
		fieldtype: "Data",
		width: "200px",
	  },
	  {
		fieldname: "select",
		label: __("put number of row"),
		fieldtype: "Int",
		width: "200px",
		on_change: function (report) {
		  var values = report.get_values();
		  var intValue = values.select;
		  $(".dt-row").find(":input[type=checkbox]").prop("checked", false);
  
		  if (!isNaN(intValue) && intValue > 0) {
			// Start selecting from row 2 (index 1)
			$(".dt-row").each(function (index) {
			  if (index >= 2 && index < 2 + intValue) {
				// Select 10 rows starting from row 2
				$(this).find(":input[type=checkbox]").prop("checked", true);
			  }
			});
		  }
		  console.log("value", intValue);
		},
	  },
	  // You can add more filters here as per your requirement
	],
	get_datatable_options(options) {
	  return Object.assign(options, {
		checkboxColumn: true,
	  });
	},
	onload: function (report) {
	  report.page.add_inner_button(__("Get Selected"), function () {
		var values = report.get_values();
		var member = values.member;
		var committee_name = values.committee_name;
		var territory = values.territory;
		var custom_customer_status = values.custom_customer_status;
		var name_of_the_cioowner_of_the_company =
		  values.name_of_the_cioowner_of_the_company;
		var selected_rows = [];
		$(".dt-scrollable")
		  .find(":input[type=checkbox]")
		  .each((idx, checkbox) => {
			if (checkbox.checked) {
			  var rowData = $(checkbox)
				.closest(".dt-row")
				.find(".dt-cell__content")
				.map(function () {
				  return $(this).text().trim();
				})
				.get();
			  var member_code = rowData[2];
			  var ceo_name = rowData[5];
			  var committees = rowData[6];
			  var export_value_in_egp = rowData[7];
			  var export_value_in_usd = rowData[8];
			  var export_volume = rowData[9];
			  var governorate = rowData[10];
			  var customer_status = rowData[11];
			  var ceo_mobile = rowData[12];
			  var email = rowData[13];
			  console.log("EMAI", export_volume);
			  ///////////////////////
			  selected_rows.push({
				member_code: member_code,
				ceo_name: ceo_name,
				committees: committees,
				export_value_in_egp: export_value_in_egp,
				export_value_in_usd: export_value_in_usd,
				export_volume: export_volume,
				governorate: governorate,
				customer_status: customer_status,
				ceo_mobile: ceo_mobile,
				email: email,
			  });
			  //////////////////////////
			}
		  });
		// for (let row of selected_rows) {
		frappe.call({
		  method: "frappe.client.insert",
		  args: {
			doc: {
			  doctype: "Foreigner  Affairs  Opportunities",
			  member_code: member,
			  ceo_name: name_of_the_cioowner_of_the_company,
			  committee: committee_name,
			  governorate: territory,
			  customer_status: custom_customer_status,
			  member_info: selected_rows,
			},
		  },
		  callback: function (r) {
			console.log(r.message);
			console.log(r.message.name);
			frappe.set_route(
			  "Form",
			  "Foreigner  Affairs  Opportunities",
			  r.message.name
			);
		  },
		});
  
		//}
		/////////////
	  });
	},
  };